libraries/chain/system_calls.cpp

```diff
@@ -387,17 +427,48 @@ THUNK_DEFINE( void, apply_upload_contract_operation, ((const protocol::upload_co
 
    context.resource_meter().use_compute_bandwidth( compute_load::medium );
 
-   auto tx_id       = crypto::hash( crypto::multicodec::sha2_256, context.get_transaction().header() );
-   auto sig_account = system_call::recover_public_key( context, system_call::get_transaction_signature( context ), util::converter::as< std::string >( tx_id ) );
+   auto contract_meta_object = system_call::get_object( context, state::space::contract_metadata(), o.contract_id() );
+   bool authorize_override = false;
+
+   if ( contract_meta_object.exists() )
+   {
+      auto contract_meta = util::converter::to< contract_metadata_object >( contract_meta_object.value() );
+      authorize_override = contract_meta.authorizes_call_contract();
+   }
+
+   bool authorized = false;
+
+   if ( authorize_override )
+   {
+      authorize_arguments args;
+      args.set_type( authorization_type::call_contract );
+
+      authorized = authorize( context, o.contract_id(), args );
```

Comment:
It would be good to be able to pass here in `authorize()` the hash of the bytecode. So the previous contract can evaluate if this is the desired code to be uploaded.
For instance, when a DAO wants to update its own contract: They could vote and save in a variable the hash of the new contract. Then when `apply_upload_contract` is called, the previous contract will check if the hash matches and authorizes the update.

Reply:
I saw you are implementing the `get_transaction` thunk 👍 which gives more flexibility. So I think what I was proposing can be covered with this new thunk 😃